### ISSUE TITLE
Unique filenames for downloaded images

### ISSUE THREAD

**Issue by jakem72360** [vector-im/element-web#19155]
### Your use case

When other users send images from their clipboards (or by drag and drop), the files are given generic names such as `image.jpg` or `image.png`. This presents a problem for the receiving user(s) when downloading the image within Element, as the user is prompted to either overwrite an existing file with the same name, or give a unique name to the image.

This can be quite tedious, and effectively defeats the benefit of the easy-to-access download button next to each message, since the user has to stop and give a unique filename to each downloaded image.

I propose having Element automatically name files based on their Matrix Content IDs (`mxc://`) when downloaded, if they would otherwise have an `image.*` filename.

**Comment by gsouquet** [comment]
When testing this, it appears to happen only in unencrypted.
Encrypted rooms set the correct filename that was given when uploading the media. 

Will reclassify as a defect

**Comment by t3chguy** [comment]
> When testing this, it appears to happen only in unencrypted.
Encrypted rooms set the correct filename that was given when uploading the media.

No, the issue is specifically when pasting an image buffer from your clipboard, there the Browser provides the file name as `image.png` 

![image](https://user-images.githubusercontent.com/2403652/135067820-b3e18681-4738-4412-b76d-912a07d4539e.png)

The suggestion for enhancement in this issue is to generate a better name/at download use the mxc hash instead.

**Comment by nidico** [comment]
I'd suggest that the generated file name shall include the upload date (!) and possibly also `matrix` and room name.

**Comment by ShadowJonathan** [comment]
> I'd suggest that the generated file name shall include the upload date (!) and possibly also `matrix` and room name.

I dont think that's a good idea wrt privacy.

---

I propose a different solution, as matrix currently also allows including the file name details when downloading images (through info on the event that references it), i recommend the following template for downloading unencrypted images;

`{mxc_domain}-{mxc_id}-{filename}.{ext}`

And the following for encrypted ones;

`{mxc_id}-{filename}.{ext}`

(omit filename and preceding dash if its not provided)

This'd provide a decent-enough tradeoff between privacy and uniqueness of the image filenames.

**Comment by nidico** [comment]
Right, including the matrix room might be a bit privacy invasive in some cases. I still think that including the upload or download datetime would be very useful. This is what Wire does and I appreciate to have at least this level of usable information in the filename.
